sql/src/main/java/org/apache/druid/sql/calcite/run/EngineFeature.java

```diff
@@ -20,12 +20,18 @@
 package org.apache.druid.sql.calcite.run;
 
 import org.apache.druid.sql.calcite.external.ExternalDataSource;
+import org.apache.druid.sql.calcite.planner.PlannerContext;
 
 /**
- * Arguments to {@link QueryFeatureInspector#feature(QueryFeature)}.
+ * Arguments to {@link EngineFeatureInspector#feature(EngineFeature, PlannerContext)}.
  */
-public enum QueryFeature
+public enum EngineFeature
 {
+  /**
+   * Can execute INSERT and REPLACE statements.
+   */
+  CAN_INSERT,
```

Comment:
This one is shorter? 🙂

I could go either way on this one to be honest. I don't have a super strong argument for one vs. the other. For that reason I'm inclined to leave it.

Reply:
The granular-by-statement-type is really what I had in mind here. I just also made two unspoken assumptions:

- All engines support SELECT so there is no feature for it
- All engines support INSERT if and only if they also support REPLACE, so there is no need for CAN_INSERT and CAN_REPLACE to be different

I suppose it's fine to get rid of these assumptions and go with CAN_SELECT, CAN_INSERT, CAN_REPLACE. I was trying to keep it neat and simple, but maybe I failed, since it generated a bunch of comments!

Reply:
By popular demand I've elected to go with more explicit choices: CAN_SELECT, CAN_INSERT, and CAN_REPLACE. I've also made the error messages nicer. Now, errors about statements like this look like:

```
Cannot execute INSERT with SQL engine 'native'.
```